engine/src/cgimageutil.cpp

```diff
@@ -180,17 +205,84 @@ bool MCGRasterToCGImage(const MCGRaster &p_raster, MCGRectangle p_src_rect, CGCo
 
 ////////////////////////////////////////////////////////////////////////////////
 
-bool MCGImageToCGImage(MCGImageRef p_src, MCGRectangle p_src_rect, CGColorSpaceRef p_colorspace, bool p_copy, bool p_invert, CGImageRef &r_image)
+void MCGImageDataProviderReleaseDataCallback(void *p_info, const void *p_data, size_t p_size)
+{
+	MCGImageRef t_image;
+	t_image = (MCGImageRef)p_info;
+	
+	MCGImageRelease(t_image);
+}
+
+bool MCGImageCreateCGDataProvider(MCGImageRef p_src, const MCGIntegerRectangle &p_src_rect, CGDataProviderRef &r_data_provider)
 {
+	MCAssert(p_src_rect.origin.x >= 0 && p_src_rect.origin.y >= 0);
+	MCAssert(p_src_rect.origin.x + p_src_rect.size.width <= (uint32_t)MCGImageGetWidth(p_src));
+	MCAssert(p_src_rect.origin.y + p_src_rect.size.height <= (uint32_t)MCGImageGetHeight(p_src));
+	
+	bool t_success = true;
+	
 	MCGRaster t_raster;
+	if (t_success)
+		t_success = MCGImageGetRaster(p_src, t_raster);
+	
+	CGDataProviderRef t_data_provider;
+	t_data_provider = nil;
+	
+	if (t_success)
+	{
+		const void *t_src_ptr;
+		t_src_ptr = MCGRasterGetPixelPtr(t_raster, p_src_rect.origin.x, p_src_rect.origin.y);
+		
+		t_success = nil != (t_data_provider = CGDataProviderCreateWithData(p_src, t_src_ptr, p_src_rect.size.height * t_raster.stride, MCGImageDataProviderReleaseDataCallback));
+	}
 	
+	if (t_success)
+	{
+		MCGImageRetain(p_src);
+		r_data_provider = t_data_provider;
+	}
+	
+	return t_success;
+}
+
+bool MCGImageToCGImage(MCGImageRef p_src, const MCGIntegerRectangle &p_src_rect, CGColorSpaceRef p_colorspace, bool p_invert, CGImageRef &r_image)
+{
+	MCGRaster t_raster;
 	if (!MCGImageGetRaster(p_src, t_raster))
 		return false;
 	
-	return MCGRasterToCGImage(t_raster, p_src_rect, p_colorspace, p_copy, p_invert, r_image);
+	if (p_invert)
+	{
+		return MCGRasterToCGImage(t_raster, p_src_rect, p_colorspace, true, true, r_image);
+	}
+	
+	// If we don't need to modify the data then create image with data provider that references the MCGImageRef
+	bool t_success = true;
+	
+	CGImageRef t_image = nil;
+	CGDataProviderRef t_data_provider = nil;
+	
+	if (t_success)
+		t_success = MCGImageCreateCGDataProvider(p_src, p_src_rect, t_data_provider);
+	
+	bool t_alpha;
+	t_alpha = !MCGImageIsOpaque(p_src);
+	
+	CGBitmapInfo t_bm_info;
+	t_bm_info = MCGPixelFormatToCGBitmapInfo(kMCGPixelFormatNative, t_alpha);
+	
+	if (t_success)
+		t_success = nil != (t_image = CGImageCreate(p_src_rect.size.width, p_src_rect.size.height, 8, 32, t_raster.stride, p_colorspace, t_bm_info, t_data_provider, nil, true, kCGRenderingIntentDefault));
```

Comment:
@livecodeian Please don't add code with nested assignments like this.  It's difficult to read, even with a good knowledge of C++ operator precedence, and goes against recommend C++ style best practice.